channeldb/migration_xx_invoices.go

```diff
@@ -0,0 +1,153 @@
+package channeldb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/lightningnetwork/lnd/zpay32"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// migrateInvoices adds invoice htlcs and a separate cltv delta field to the
+// invoices.
+func migrateInvoices(tx *bbolt.Tx) error {
+	log.Infof("Migrating invoices to new invoice format")
+
+	invoiceB := tx.Bucket(invoiceBucket)
+	if invoiceB == nil {
+		return ErrNoInvoicesCreated
+	}
+
+	// Iterate through the entire key space of the top-level invoice bucket.
+	// If key with a non-nil value stores the next invoice ID which maps to
+	// the corresponding invoice. Store those keys first, because it isn't
+	// safe to modify the bucket inside a ForEach loop.
+	var invoiceKeys [][]byte
+	err := invoiceB.ForEach(func(k, v []byte) error {
+		if v == nil {
+			return nil
+		}
+
+		invoiceKeys = append(invoiceKeys, k)
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	nets := []*chaincfg.Params{
+		&chaincfg.MainNetParams, &chaincfg.SimNetParams,
+		&chaincfg.RegressionNetParams, &chaincfg.TestNet3Params,
+	}
+
+	// Iterate over all stored keys and migrate the invoices.
+	for _, k := range invoiceKeys {
+		v := invoiceB.Get(k)
+
+		// Deserialize the invoice with the deserializing function that
+		// was in use for this version of the database.
+		invoiceReader := bytes.NewReader(v)
+		invoice, err := deserializeInvoiceLegacy(invoiceReader)
+		if err != nil {
+			return err
+		}
+
+		// Try to decode the payment request for every possible net to
+		// avoid passing a the active network to channeldb. This would
+		// be a layering violation, while this migration is only running
+		// once and will likely be removed in the future.
+		var payReq *zpay32.Invoice
+		for _, net := range nets {
```

Comment:
Heh, not the prettiest logic...but it works :p

Reply:
Yes, we can't be perfect...